### PR TITLE
[master] Apple Silicon support

### PR DESCRIPTION
**Pull request by shahmishal**

### CODE REVIEW

**Comment by shahmishal** [comment]
@swift-ci test

**Comment by shahmishal** [comment]
Screen Shot 2020-07-03 at 5 53 19 PM

```
swift/release/5.3 -> swift/master
- 12 unmerged commits. 1 commits are currently being merged/build/tested.
- Unmerged commits:
* 8e65afd5a004d6ee3b2420efba17d053775c1c52 Fri Jul 3 10:04:43 2020: Conflict
... 10 commits in-between ...
* f0497aca6e6f7945ca429cdcb03aa961c0335b5d Wed Jul 1 19:30:25 2020: Auto merge in progress
```

The bot failed because automerger is currently blocked.

File: include/swift/AST/AvailabilitySpec.h [hunk 1]
```diff
@@ -68,16 +68,33 @@ class PlatformVersionConstraintAvailabilitySpec : public AvailabilitySpec {
   SourceLoc PlatformLoc;
 
   llvm::VersionTuple Version;
+
+  // For macOS Big Sur, we we canonicalize 10.16 to 11.0 for compile-time
```
Comment (AliSoftware) [inline include/swift/AST/AvailabilitySpec.h:72]:
```suggestion
  // For macOS Big Sur, we canonicalize 10.16 to 11.0 for compile-time
```

**Review by DougGregor — APPROVED** [review]
Looks great, thank you!

**Comment by shahmishal** [comment]
@swift-ci test

File: cmake/modules/DarwinSDKs.cmake [hunk 2]
```diff
@@ -4,17 +4,34 @@ option(SWIFT_ENABLE_IOS32
...
+
+# Remove the last component, if any. e.g. 10.15.26 -> 10.15
+string(REGEX REPLACE "\([0-9]*[.][0-9]*\)[.][0-9]*" "\\1"
+       SWIFT_OSX_SDK_VERSION "${SWIFT_OSX_SDK_VERSION}")
+
+if (${SWIFT_OSX_SDK_VERSION} STREQUAL "10.14" OR
+    ${SWIFT_OSX_SDK_VERSION} STREQUAL "10.15")
+  set(SUPPORTED_OSX_ARCHS "x86_64")
+else()
+  set(SUPPORTED_OSX_ARCHS "x86_64;arm64e")
```
Comment (cltnschlosser) [inline cmake/modules/DarwinSDKs.cmake:33]: Should this include arm64 like line 17?

File: lib/AST/PlatformKind.cpp [hunk 3]
```diff
@@ -155,3 +156,17 @@ bool swift::inheritsAvailabilityFromPlatform(PlatformKind Child,
 
   return false;
 }
+
+llvm::VersionTuple swift::canonicalizePlatformVersion(
+    PlatformKind platform, const llvm::VersionTuple &version) {
+
+  // Canonicalize macOS version for macOS Big Sur to great
```
Comment (xwu) [inline lib/AST/PlatformKind.cpp:163]:
```suggestion
  // Canonicalize macOS version for macOS Big Sur to treat
```

File: lib/Parse/ParseExpr.cpp [hunk 4]
```diff
@@ -3812,6 +3812,11 @@ Parser::parsePlatformVersionConstraintSpec() {
   // Register the platform name as a keyword token.
   TokReceiver->registerTokenKindChange(PlatformLoc, tok::contextual_keyword);
 
+  // Keep the original version around for run-time checks to support
+  // macOS Big Sur betas that report 10.16 at
+  // at run time.
```
Comment (xwu) [inline lib/Parse/ParseExpr.cpp:3817]:
```suggestion
  // run time.
```

File: cmake/modules/DarwinSDKs.cmake (hunk 2 above)
Comment (shahmishal) [reply cmake/modules/DarwinSDKs.cmake:33]: Going to file a bugs.swift.org and follow up in the future PR.

**Comment by shahmishal** [comment]
Full test was already done, going to only run smoke test for comment update. 
@swift-ci smoke test

**Comment by shahmishal** [comment]
@swift-ci smoke test

**Comment by shahmishal** [comment]
Full test results: 
✅ macOS - https://ci.swift.org/job/swift-PR-osx/22084/
✅ Linux - https://ci.swift.org/job/swift-PR-Linux/22145/
✅ Windows - https://ci-external.swift.org/job/swift-PR-windows/4822/

File: cmake/modules/DarwinSDKs.cmake (hunk 2 above)
Comment (DougGregor) [reply cmake/modules/DarwinSDKs.cmake:33]: This bit of code should be dead now

**Pull request merged**